syncany-lib/src/main/java/org/syncany/database/dao/FileHistorySqlDao.java

```diff
@@ -166,6 +168,111 @@ public FileHistoryId expandFileHistoryId(FileHistoryId fileHistoryIdPrefix) {
+			// This first query retrieves the last version for each FileHistory matching the three requested properties.
+			// However, it does not guarantee that this version is indeed the last version in that particular
+			// FileHistory, so we need another query to verify that.
+
+			preparedStatement.setString(1, filecontentChecksum);
+			preparedStatement.setLong(2, size);
+			preparedStatement.setTimestamp(3, new Timestamp(modifiedDate.getTime()));
+
+			try (ResultSet resultSet = preparedStatement.executeQuery()) {
+				Collection<PartialFileHistory> fileHistories = new ArrayList<>();
+				while (resultSet.next()) {
+					String fileHistoryId = resultSet.getString("filehistory_id");
+					PartialFileHistory fileHistory = getLastVersionByFileHistoryId(fileHistoryId);
+					boolean resultIsLatestVersion = fileHistory.getLastVersion().getVersion() == resultSet.getLong("version");
+					boolean resultIsNotDelete = fileHistory.getLastVersion().getStatus() != FileVersion.FileStatus.DELETED;
+
+					// Only if the result is indeed the last in it's history, we can use it
+					// to base other versions off it. So we return it.
+					if (resultIsLatestVersion && resultIsNotDelete) {
+						fileHistories.add(fileHistory);
+					}
+				}
+				return fileHistories;
+			}
+
+		}
+		catch (SQLException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	/**
+	 * This function returns a FileHistory, with as last version a FileVersion with
+	 * the given path. 
+	 * 
+	 * If the last FileVersion referring to this path is not the last in the
+	 * FileHistory, or no such FileVersion exists, null is returned.
+	 */
+	public PartialFileHistory getFileHistoryWithLastVersionByPath(String path) {
+		try (PreparedStatement preparedStatement = getStatement("filehistory.select.master.findLatestFileVersionsForPath.sql")) {
+			preparedStatement.setString(1, path);
+			try (ResultSet resultSet = preparedStatement.executeQuery()) {
+				// Fetch the latest versions of all files that once existed with the given path and find the most recent by comparing vector clocks
+				String latestFileHistoryId = null;
+				Long latestFileVersion = null;
+				VectorClock latestVectorClock = null;
+				while (resultSet.next()) {
+					VectorClock clock = VectorClock.parseVectorClock(resultSet.getString("vectorclock_serialized"));
+					if (latestVectorClock == null || VectorClock.compare(clock, latestVectorClock) == VectorClock.VectorClockComparison.GREATER) {
+						latestVectorClock = clock;
+						latestFileHistoryId = resultSet.getString("filehistory_id");
+						latestFileVersion = resultSet.getLong("version");
+					}
+				}
+
+				// If no active file history exists for this path, return
+				if (latestFileHistoryId == null) {
+					return null;
+				}
+
+				// Find for the latest file history the last file version to check if the file has moved
+				PartialFileHistory fileHistory = getLastVersionByFileHistoryId(latestFileHistoryId);
```

Comment:
That comment made way more sense in context with how I thought about it with Tim. I've adjusted it to make sense out of context. Is it understandable now?